### ISSUE TITLE
Feature request: save advance pull/fetch options

### ISSUE THREAD

**Issue by lipkau** [gitextensions/gitextensions#4895]
## Current Behavior

As a user, I am able to perform the predefined actions and additionally define the default behavior of the toolbar button.
![image](https://user-images.githubusercontent.com/4992870/39386612-c6e504d2-4a75-11e8-8ca0-807a8c788fd1.png)

## Desired Behavior

Personally, I would like to be able to have `Fetch all` to always `prune remote branches`.
I have tried to combine `Set the next selection as default` with `Open pull dialog` - unfortunately, did not the trick.

## Poposition

I don't have the perfect solution, but here are some ideas:

* Have an entry in `Commands` to prune all
* Be able to set the checkbox for prune persistently

**Comment by RussKie** [comment]
This can be achieved with custom `AfterPull` scripts (Settings -> Scripts), however looking at the code these scripts will only get triggered via the Pull dialog and not for the pull or fetch commands in the toolbar (which is a bug in itself).

**Comment by RussKie** [comment]
Related: #4909 and  #4910

**Comment by zvirja** [comment]
@RussKie This is a really missing feature. I was reading around about the suggested way to solve the issue and feel that it will not provide the required level of flexibility. I don't need to run `prune` each time I perform a fetch. Rather, I would like to control when I want fetch with prune and when without. If I write a script, it will run _for each time_.

I suggest to add two more buttons to the pop-up menu - "Fetch and prune" and "Fetch and prune all". The icon should be also adjusted, to have e.g. red cross to warn. Click on any button will run the fetch with the appropriate flag set.

If you wish, we could even disable that button by default and enable it via settings.

What do you think? I'll fire a PR for that if you are fine 😊

**Comment by RussKie** [comment]
Could you please clarify the difference between "_Fetch and prune_" and "_Fetch and prune all_"?

Overall I'm really ambivalent, @gitextensions/git-extensions-source thoughts?

**Comment by vbjay** [comment]
Look at the pull dialog.

**Comment by zvirja** [comment]
Hi guys. I decided to proceed with the `Fetch and prune all` only to avoid the confusion (as you pointed out). Already implemented that in branch, here is how it looks:
![image](https://user-images.githubusercontent.com/1074182/43716731-e655df5a-998d-11e8-999f-8f027c9ed9be.png)

I made the icon red, to reflect the `prune` aspect.

Are you fine with how it looks?

**Comment by RussKie** [comment]
LGTM.
We'll need to have some docs for it

**Comment by vbjay** [comment]
Did you look at the pull dialog?

**Comment by vbjay** [comment]
Frankly that has NO business being on a menu. Too easy to hit it by mistake.

**Comment by zvirja** [comment]
> Did you look at the pull dialog?

What do you exactly mean? I saw the dialog and the ability to tick the `prune` checkbox. The problem is that it's pretty inconvenient and annoying, as I need to perform a set of clicks:
- click to open the dialog
- select _All_ remotes
- Tick the _Prune_ checkbox
- Tick the pull button

All that instead of single click. And I need that very often 😟 

> Frankly that has NO business being on a menu.

If we look from that perspective, there is no business for each button on that strip - all the items are just shortcuts for the actions we achievable from the Pull Dialog 😉 It's a matter of convenience only, so we optimize the most common scenarios.

> Too easy to hit it by mistake.

What if I make it as a sub-menu for the `Fetch all` item:
![image](https://user-images.githubusercontent.com/1074182/43718208-a152af0a-9992-11e8-92de-9a0ca32be70b.png)

This way chances to hit it accidentally are much less. But it might be non-obvious that `Fetch all` is clickable as well 😕 

Alternatively we might disable that button by default and enable it via advanced setting only. This way only interested and aware people enable it. I'd vote for this option - make the button flat and disabled by default (as it's indeed a dangerous operation).

**Comment by vbjay** [comment]
The dialog handles it just fine and prevents accidents. If you really
desire the functionality, create a git alias that allows you to pass in the
remote and call it in the console.

**Comment by RussKie** [comment]
A second menu item is ok, 2nd level menu makes it confusing like you said

**Comment by vbjay** [comment]
> A second menu item is ok, 2nd level menu makes it confusing like you said

I STRONGLY!!!!!! disagree.  You try having fine motor issues and accidently clicking a destructive menu.  This is not the way to go.  Whining about a dialog being too much when it does the job of preventing a destructive action and makes the user jump through hoops because IT IS destructive.  Set an alias and be done.

**Comment by vbjay** [comment]
https://git-scm.com/book/en/v2/Git-Basics-Git-Aliases

**Comment by vbjay** [comment]
@jbialobr @KindDragon @EbenZhang Please chime in on my point.

**Comment by vbjay** [comment]
![image](https://user-images.githubusercontent.com/3629489/43723410-96a5effe-9965-11e8-8464-672ade3b10d4.png)
Yes.  You can.  Just add a script with the add to revision grid option checked.  If you want to shorten the command, add an alias and call the alias.  
![image](https://user-images.githubusercontent.com/3629489/43723564-eff82a68-9965-11e8-9b7b-b6fce0d911e8.png)

**Comment by drewnoakes** [comment]
@zvirja, is there a case where you would fetch without pruning remote refs? Perhaps we could add a setting that adds `--prune` to `fetch` and `pull` operations, without otherwise changing the UI.

**Comment by vbjay** [comment]
UI already has it.  
![image](https://user-images.githubusercontent.com/3629489/43725254-08c2d63e-996a-11e8-91de-19194bb4fb6d.png)

If he adds a script to the grid, with git fetch {UserInput} --prune he can

1. Right click grid
2. Click script menu
3. type remote name
4. Press enter

**Comment by vbjay** [comment]
If doing all remotes, it's even easier and you don't need the user input argument.

**Comment by EbenZhang** [comment]
Personally I have been using the --prune for about 4 years and never been bothered. I would vote for the additional non-sub menu.

**Comment by jbialobr** [comment]
I second the @drewnoakes question as I don't know a reason why not to always prune. In case there is a good reason to not always prune then I vote for adding another 1st level menu with confirmation which can be suppressed in the advanced settings.

**Comment by zvirja** [comment]
@vbjay Many thanks for the advice 🙏 Indeed, the script option allows me to shortcut the action, which is awesome! 😊 💃 

Now the issue isn't that critical to me, thanks for that! However, I'd still desire to have the `Fetch & prune all` button, as it provides advances:
- I don't need to configure tool on each machine
- It has more logical placement
- The experience is more aligned (e.g. I don't have pop-up console)

Therefore, if you could agree among yourselves, I could fire a PR with button and confirmation which can be disabled ☺️ 

> Perhaps we could add a setting that adds --prune to fetch and pull operations, without otherwise changing the UI.

It might indeed satisfy my needs, however I'd vote to not do that. It's very implicit and you should remember that clicking the `Fetch all` you actually prune the branches (same as using `git config --global fetch.prune true` suggested by @EbenZhang) as you enabled that somewhere deeply inside. The standalone button is more explicit with that regards.

**Comment by vbjay** [comment]
As long as it can be removed or disabled, I'm fine.

**Comment by zvirja** [comment]
@vbjay The idea is to have the button always available. However when you click on it, you'll get a confirmation pop-up. That will prevent the accidental run you worried about. Are you happy with that?

Could you please also describe why you consider prune as something dangerous (as it seems you are the only guy who sees that)? It would be quite interesting to understand your point.

Thanks.

**Comment by vbjay** [comment]
Prune removes tracking branches when a branch has been removed on the
remote. I have fine motor issues from a stroke when I was an infant. I have
clicked menus next to the one I want TOO many times.  Menus that are in any
way destructive no matter how minor should not fire off the action unless
there is a confirmation or I can disable it.

If you have confirmation put ... at the end of the text.

**Comment by zvirja** [comment]
> Prune removes tracking branches when a branch has been removed on the remote. 

Can you please elaborate on this behavior? I never saw it. Just tried:
- I have branches as here:
    ```
    $ git branch -vvv -a
    * master                                            6c67c9bf [upstream/master] Fix Pluralsight course link on README (#1063)
      test-branch                                       6c67c9bf [origin/test-branch] Fix Pluralsight course link on README (#1063)
      remotes/origin/master                             6c67c9bf Fix Pluralsight course link on README (#1063)
      remotes/origin/test-branch                        6c67c9bf Fix Pluralsight course link on README (#1063)
      remotes/upstream/master                           6c67c9bf Fix Pluralsight course link on README (#1063)
    ```

- I checkout `master` to make experiment clear
- I remove branch using GitHub UI
- I run `git fetch --all --prune`:
    ```
    $ git fetch --all --prune
    Fetching upstream
    Fetching origin
    From github.com:zvirja/AutoFixture
     - [deleted]           (none)     -> origin/test-branch
    ```
- I check branches again:
    ```
    $ git branch -vvv -a
    * master                                            6c67c9bf [upstream/master] Fix Pluralsight course link on README (#1063)
      test-branch                                       6c67c9bf [origin/test-branch: gone] Fix Pluralsight course link on README (#1063)
      remotes/origin/master                             6c67c9bf Fix Pluralsight course link on README (#1063)
      remotes/upstream/master                           6c67c9bf Fix Pluralsight course link on README (#1063)
    ```

As expected, the remote-tracking reference (`remotes/origin/test-branch`) was removed, while the local branch (`test-branch`) is still here. That's exactly why I don't treat `prune` as a dangerous operation.

Could you please provide the scenario where it's destructive?

> If you have confirmation put ... at the end of the text.

Not sure I got you, please clarify.

**Comment by vbjay** [comment]
https://docs.microsoft.com/en-us/windows/desktop/uxguide/cmd-menus#using-ellipses
.

I didn't say dangerous. I said destructive. It's possible to have a
tracking branch and not a local branch. Yes I know you can recreate
branches. I prefer to run a prune as a specific action and I do not want to
clear out tracking branches till I have verified. I also don't want a menu
action to cause me to clear out tracking when I didn't want to.
